Adding minute_file_sink_mt #2727

mmanoj opened the pull request:
Adding minute_file_sink_mt #2727 to support minute_file_sink to support every 1 min file rotation.
[minute_file_sink.h.txt](https://github.com/gabime/spdlog/files/11395573/minute_file_sink.h.txt)


https://github.com/gabime/spdlog/commit/5306b1fd58b84a8b8471c1e0efb6750614a66002

mmanoj committed:
minute_file_sink to support every 1 min file rotation


example driver code:
auto logger = spdlog::minute_logger_mt("minutes_basic_logger", "logs/min-log.txt",false,60);

mmanoj committed:
Merge branch 'gabime:v1.x' into v1.x

tt4g commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,208 @@
+    {
+        filename_t basename, ext;
+        std::tie(basename, ext) = details::file_helper::split_by_extension(filename);
+        return fmt_lib::format(SPDLOG_FILENAME_T("{}_{:04d}-{:02d}-{:02d}-{:02d}_{:02d}{}"), basename, now_tm.tm_year + 1900, now_tm.tm_mon + 1,
```
```suggestion
        return fmt_lib::format(SPDLOG_FMT_STRING(SPDLOG_FILENAME_T("{}_{:04d}-{:02d}-{:02d}-{:02d}_{:02d}{}")), basename, now_tm.tm_year + 1900, now_tm.tm_mon + 1,
```

mmanoj replied:
Updated, under testing

tt4g commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,208 @@
+        filename_t basename, ext;
+        std::tie(basename, ext) = details::file_helper::split_by_extension(filename);
+        return fmt_lib::format(SPDLOG_FILENAME_T("{}_{:04d}-{:02d}-{:02d}-{:02d}_{:02d}{}"), basename, now_tm.tm_year + 1900, now_tm.tm_mon + 1,
+            now_tm.tm_mday, now_tm.tm_hour,now_tm.tm_min,ext);//colock details define in os-inl.h
```
```suggestion
            now_tm.tm_mday, now_tm.tm_hour,now_tm.tm_min,ext);
```

tt4g commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,208 @@
+class minute_file_sink final : public base_sink<Mutex>
+{
+public:
+    // create hourly file sink which rotates on given time
```
```suggestion
    // create every minute file sink which rotates on given time
```

tt4g commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,208 @@
+        bool should_rotate = time >= rotation_tp_;
+        if (should_rotate)
+        {
+            if (remove_init_file_)
```
I don't understand why you would delete an empty log file.
Why is this necessary?

mmanoj replied:
This part kept as it is from hourly log sink.

`bool should_rotate = time >= rotation_tp_;
        if (should_rotate)
        {
            if (remove_init_file_)
            {
                file_helper_.close();
                details::os::remove(file_helper_.filename());`

tt4g replied:
Somehow I figured out why this code is there.
Perhaps, the author of sink wanted to clean up the log file created in the constructor, if it was empty, removing it at the time of rotation.

The following scenario is assumed.

1. Create log file `log.2023-01-01-12-01.txt` in the constructor.
2. More than one minute passes without a log being recorded.
3. The contents of the file to be rotated are empty. Delete it as unnecessary and record log messages in a new log file.

Can you add a comment to this code as I am having difficulty understanding the purpose of this code?

tt4g commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,208 @@
+ * If max_files > 0, retain only the last max_files and delete previous.
+ */
+template<typename Mutex, typename FileNameCalc = minute_filename_calculator>
+class minute_file_sink final : public base_sink<Mutex>
```
daily_sink and hourly_sink can change the rotation time, but `minute_file_sink` seems to have a fixed rotation time. I suggest renaming it to `every_minute_file_sink` to avoid confusion.

tt4g commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,208 @@
+    {
+        auto now = log_clock::now();
+        tm date = now_tm(now);
+        date.tm_min = 0;
```
```suggestion

```

The most important rotation feature seems to be broken.
The minute of the next rotation time is always 1 (1 minute every hour).

mmanoj replied:
@tt4g 
Thanks for suggestions and review,I will add configuration to make X MIN interval to do the rotation as configurable number of minutes.Hope it more flexible.I also go through other changes as well

mmanoj replied:
Updated with configurable rotation interval minutes

auto logger = spdlog::minute_logger_mt("basic_logger", "logs/min-log.txt",false,60,2);

tt4g reviewed (commented):
I commented on the points of concern.
It must also pass @gabime review after modification.

mmanoj committed:
Review fixes

Fixes as per suggestions

mmanoj reviewed (commented):
fixed suggested changes and configurable minutes

tt4g commented:
I think the current rotation implementation is wrong.
`minute_file_sink` rotates at intervals (minutes).
However, the daily sink and hourly sink rotate at a specified time (hour or minute).
If you are adding sinks that work differently, I think you should change the name of the sinks to make it easier to understand.

mmanoj commented:
@tt4g 

Only sink name change required? hope the other implemented logic and suggestions are fixed?

tt4g commented:
> Only sink name change required? hope the other implemented logic and suggestions are fixed?

Did you not develop this sink because you wanted it?
If you wanted this sink that rotates on an interval, change the name.
If not, please fix it.

gabime commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,217 @@
+        
+        if (rotation_minute < 0 || rotation_minute > 59)
+        {
+            throw_spdlog_ex("daily_file_sink: Invalid rotation time in ctor");
```
1. Why can't it be greater than 59? what would happen if someone wants 90 minutes for example?
2. should be "minute_file_sink: Invalid rotation_minutes value in ctor"

mmanoj replied:
@gabime 

Thanks for the feedback and got your point, remove the validation considering the concern.

tt4g replied:
@mmanoj Note: In the current rotation implementation, `rotation_m_` should not be allowed to be `0`.

mmanoj replied:
@tt4g 
Implement the check for 0 min.

mmanoj commented:
@tt4g 

Yes, This is required as per my project requirements and hope many people like the feature.I change the name to "minute_interval_logger_mt" to make it meaningful.

mmanoj committed:
Add files via upload

Change the logger name to reflect the minutes interval based rotation and remove the check for max min 59 in configuration.

mmanoj committed:
Update check for 0 min configuration check.

mmanoj committed:
corrected  as rotation_minute <= 0

gabime commented:
@mmanoj  Please add a unit test for this sink. Something very basic just to verify it compiles right (and perhaps test to validate it throws on invalid input).

tt4g commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,217 @@
+
+template<typename Factory = spdlog::synchronous_factory>
+inline std::shared_ptr<logger> minute_interval_logger_st(const std::string &logger_name, const filename_t &filename, bool truncate = false,
+    uint16_t max_files = 0,int minute = 0, const file_event_handlers &event_handlers = {})
```
```suggestion
    uint16_t max_files = 0, int minute = 1, const file_event_handlers &event_handlers = {})
```

`minute = 0` will thrown.

tt4g commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,217 @@
+//
+template<typename Factory = spdlog::synchronous_factory>
+inline std::shared_ptr<logger> minute_interval_logger_mt(const std::string &logger_name, const filename_t &filename, bool truncate = false,
+    uint16_t max_files = 0, int minute = 0, const file_event_handlers &event_handlers = {})
```
```suggestion
    uint16_t max_files = 0, int minute = 1, const file_event_handlers &event_handlers = {})
```

`minute = 0` will thrown.

mmanoj commented:
> @mmanoj Please add a unit test for this sink. Something very basic just to verify it compiles right (and perhaps test to validate it throws on invalid input).

@gabime Sure,can you refer some example, I'm checking under test/test_daily_logger

mmanoj committed:
corrected as per suggestions

gabime commented:
1. Just look at any test in the tests for example. please create a new file in tests with onr or two tests.
2. Also, I think a better name for the sink would be “periodic_file_sink”. This way in the future it could be extended to support any chrono units, not just minutes.
3. Please remove the lime with the email address from the top comments. I try to have same header format for all files

mmanoj commented:
> 
1. Working On the test cases files
2. Rename the file and update the sink names as per the advice.Also we can make it  seconds as base unit, so we can represent minutes and hours from that. What do you think?
3. Done.

mmanoj committed:
Update as per comments

Updated as per advice.Please check and update if any changes required.

mmanoj commented:
https://github.com/mmanoj/spdlog/blob/v1.x/include/spdlog/sinks/periodic_file_sink.h

gabime commented:
1. I meant using `std::chrono::duration<Rep, Period> interval` as for the interval. see https://github.com/gabime/spdlog/blob/57a9fd0841f00e92b478a07fef62636d7be612a8/include/spdlog/spdlog.h#LL85C12-L85C12 for an example.

However, for this to work the generated filename should probably also contain seconds in the name and the units restricted to seconds or higher,(e.g. not accept milliseconds or micros  using  `std::enable_if`)
so I suggest leaving this for a later stage and start with minutes for this initial version

2. The name should be periodic_file_sink instead of periodic_interval_file_sink.

mmanoj committed:
using std::chrono::duration

using std::chrono::duration

mmanoj commented:
> 1. I meant using `std::chrono::duration<Rep, Period> interval` as for the interval. see https://github.com/gabime/spdlog/blob/57a9fd0841f00e92b478a07fef62636d7be612a8/include/spdlog/spdlog.h#LL85C12-L85C12 for an example.
> 
> However, for this to work the generated filename should probably also contain seconds in the name and the units restricted to seconds or higher,(e.g. not accept milliseconds or micros using `std::enable_if`) so I suggest leaving this for a later stage and start with minutes for this initial version
> 
> 2. The name should be periodic_file_sink instead of periodic_interval_file_sink.

Updated as per advice.Please check and update.Once this clear I will work on the unit test

gabime commented:
I see the pr contains 2 sink files? which one is it?

mmanoj commented:
> I see the pr contains 2 sink files? which one is it?

https://github.com/mmanoj/spdlog/blob/v1.x/include/spdlog/sinks/periodic_file_sink.h

mmanoj commented:
> > I see the pr contains 2 sink files? which one is it?
> 
> https://github.com/mmanoj/spdlog/blob/v1.x/include/spdlog/sinks/periodic_file_sink.h

@gabime 

Any feedback regarding the PR?

gabime commented on include/spdlog/sinks/minute_file_sink.h:
```diff
@@ -0,0 +1,217 @@
+// Copyright(c) 2015-present, Gabi Melman & spdlog contributors.
```
Please remove this file as we use the other one

mmanoj replied:
done

gabime commented on include/spdlog/sinks/periodic_file_sink.h:
```diff
@@ -0,0 +1,217 @@
+   // create every periodic file sink which rotates on given time
+    periodic_file_sink(
+        filename_t base_filename, bool truncate = false, uint16_t max_files = 0,
+        std::chrono::duration<int> rotation_period = 0, const file_event_handlers &event_handlers = {})
```
Should be 

```c++
  periodic_file_sink(
       filename_t base_filename, 
       std::chrono::minutes rotation_interval,
       uint16_t max_files max_files,
       bool truncate_first = false,       
       const file_event_handlers &event_handlers = {})

```

mmanoj replied:
Done

gabime commented on include/spdlog/sinks/periodic_file_sink.h:
```diff
@@ -0,0 +1,217 @@
+        {
+            throw_spdlog_ex("periodic_file_sink: Invalid rotation time in ctor");
+        }
+        
```
I think we should add some sanity restriction on max_files, especially since we store all the names in a buffer

```c++
  if (max_files > 2000)  
   {
         throw_spdlog_ex("periodic_file_sink: max_files param exceeds max of 2000");
   }
```

mmanoj replied:
Done

mmanoj committed:
Delete minute_file_sink.h

mmanoj committed:
Update as per comments

gabime commented:
Thanks. Would you add a test?

mmanoj commented:
> Thanks. Would you add a test?

If current implementation is fine,I will work on it.Please confirm.

gabime commented:
The thing is, I cannot know if it even complies without a minimal test. Just add minimal test that includes it and use it in the most basic way.

mmanoj commented:
> The thing is, I cannot know if it even complies without a minimal test. Just add minimal test that includes it and use it in the most basic way.

sure, will work on this and update you.

mmanoj commented:
> > The thing is, I cannot know if it even complies without a minimal test. Just add minimal test that includes it and use it in the most basic way.
> 
> sure, will work on this and update you.

Can you help to understand how spdlog test framework work.I'm not able to run the test cases except example application under spdlog/tests/example directory.

for example how to run test_daily_logger.cpp

kegechen commented on include/spdlog/sinks/periodic_file_sink.h:
```diff
@@ -0,0 +1,231 @@
+inline std::shared_ptr<logger> periodic_logger_mt(const std::string &logger_name, const filename_t &filename,std::chrono::minutes period = 1,uint16_t max_files = 0, bool truncate = false,
+     const file_event_handlers &event_handlers = {})
+{
+    return Factory::template create<sinks::periodic_file_sink_mt>(logger_name, filename,period,max_files, truncate, event_handlers);
```
```suggestion
    return Factory::template create<sinks::periodic_file_sink_mt>(logger_name, filename, period,max_files, truncate, event_handlers);
```

kegechen commented on include/spdlog/sinks/periodic_file_sink.h:
```diff
@@ -0,0 +1,231 @@
+inline std::shared_ptr<logger> periodic_logger_st(const std::string &logger_name, const filename_t &filename,std::chrono::minutes period = 1,uint16_t max_files = 0, bool truncate = false,
+     const file_event_handlers &event_handlers = {})
+{
+    return Factory::template create<sinks::periodic_file_sink_st>(logger_name, filename,period, max_files,truncate,   event_handlers);
```
```suggestion
    return Factory::template create<sinks::periodic_file_sink_st>(logger_name, filename, period, max_files,truncate, event_handlers);
```

kegechen commented on include/spdlog/sinks/periodic_file_sink.h:
```diff
@@ -0,0 +1,231 @@
+        : base_filename_(std::move(base_filename))
+        , file_helper_{event_handlers}
+        , truncate_(truncate_first)
+        , max_files_(max_files)
```
It would be great if the max file size was also configurable
like 20 * 1024 *1024 （20MB）
like [rotating_file_sink](https://github.com/gabime/spdlog/blob/2312489bdc304886b4a2eebc8f7f8dc0c4de688f/include/spdlog/sinks/rotating_file_sink.h#L47C14-L47C14)

mmanoj replied:
@kegechen 

Thanks for the suggestion,I will look in to that.I have some issues on running the test case as I mention above.If someone can help to fix it I can quickly close the PR.

kegechen replied:
Can't help, unfortunately I'm not Not familiar with the test case either
[Catch2 tutorial](https://github.com/catchorg/Catch2/blob/devel/docs/tutorial.md#top) may help

kegechen replied:
by the way, to many commits in your pr. should have rebase first
the title of pr also should change

mmanoj replied:
@kegechen 

Sorry for the delay due to some urgent tasks,I'm going to start working on this again.Will update the progress
